### DISCUSSION TITLE
Date and time format

### DISCUSSION THREAD

**Discussion by RedBlob** [Budibase/budibase#10730] [Help] [answered]
Hello,
I am using PostgreSQL as a datasource and I'm using a Table block to show the table data.
The table block is directly connected with a table on PostgreSQL.

How is possible to change the Date/Time format in the column ?

Thank You

![Immagine 2023-05-29 145054](https://github.com/Budibase/budibase/assets/70526244/aa2d54d3-bedb-446b-bcc8-3d8315b2c0f3)
![Immagine 2023-05-29 145759](https://github.com/Budibase/budibase/assets/70526244/414f6687-cb8c-4212-bcf2-a0a6fe6ce1dc)

**Comment by ConorWebb96** [comment, marked as answer]
Hey @RedBlob,

I've setup a quick postgres table. The data related to date/time I'm working with outputs as: `2023-05-30T10:00:00.000Z`. From here you can click on the lighting bolt shown in your screenshot and switch from `Handlebars` to `Javascript` and then write some JS to convert the date/time into the format which you desire in my case it will show like this: `dd-mm-yy h:m`.

**Before changing the format**
<img width="176" alt="Screenshot 2023-05-30 at 07 55 35" src="https://github.com/Budibase/budibase/assets/126772285/eddcb88c-9755-47f8-aa3e-52da9f69d2b4">

**js code used to change the date/time format**
```
const dateString = $("Value");
const date = new Date(dateString);
const day = String(date.getDate()).padStart(2, '0');
const month = String(date.getMonth() + 1).padStart(2, '0');
const year = String(date.getFullYear()).slice(2);
const hours = String(date.getHours()).padStart(2, '0');
const minutes = String(date.getMinutes()).padStart(2, '0');
  
return `${day}-${month}-${year} ${hours}:${minutes}`;
```

**End result of the above code**
<img width="197" alt="Screenshot 2023-05-30 at 07 56 44" src="https://github.com/Budibase/budibase/assets/126772285/bae994fd-e6e8-4fcc-9a8e-05a9b91c1a5f">

**Reply by RedBlob** [reply]
Hi @ConorWebb96 ,

Thank You very much for help ..... it works OK.

Best Regards

**Comment by gallogabor** [comment]
Hi @ConorWebb96!

I tried this method to format date fields in a table, but the JS code is not saved at all.
If insert this code in value field settings form, then press save on all open forms (in the app editor of course), the table not changing its contents (no date formattind happens), and when I go back to field value setting screen, it's empty, no JS script.

I'm running 2.7.11 (latest) version self-hosted.

**Reply by ConorWebb96** [reply]
Hey @gallogabor,

- What type of data source are you using? 
- Can you supply some screenshots of your field setup? JS code, field type, screenshot of your bindings area.

Thanks ahead of time.

**Reply by gallogabor** [reply]
Hi @ConorWebb96 

Of course, here it is:

> What type of data source are you using?

External MySQL database (MariaDB 10.3.38), InnoDB tables.
"esemely_idopontja" field is a MySQL datetime type field, NOT NULL, no index on it. All rows contain valid date and time information.

Maybe interesing in this relation that BB always displays "Invalid time value" message on row update screen for this table in app editor and in running app, too. The date and time shown by BB is correct.
Other DB clients (DBeaver desktop client and phpMyAdmin web client) has no problem with field contents.

> Can you supply some screenshots of your field setup? JS code, field type, screenshot of your bindings area.

I go to: select Table (component), Configure columns, first column (MySQL datetime field type in DB table) gear icon, Value field lightning icon, Javascript tab, paste above code (first picture), press Save (second picture), press Save again (close Configure columns editor), nothing changes on date field display, going back to Configure columns, first column gear icon (third picture), Value field empty.

Insering code
![Képernyőkép – 2023-06-13 19-14-42](https://github.com/Budibase/budibase/assets/17951393/b82d20e3-e333-456e-a77f-778e1fcbc5fb)

Inserted code saved
![Képernyőkép – 2023-06-13 19-15-00](https://github.com/Budibase/budibase/assets/17951393/410e3418-fcc6-4af9-a60f-fc8acd9594e0)

Going back to field editor after no chage in date display
![Képernyőkép – 2023-06-13 19-15-24](https://github.com/Budibase/budibase/assets/17951393/ea2445d3-1927-4e39-9b06-d03366eb837e)

**Reply by ConorWebb96** [reply]
Hey @gallogabor,

I've tried replicating this and it worked fine on my own MySQL instance with the same code provided above. Could you supply an app export and a db dump or the schema so I can have a look. Could you send them to support@budibase.com, if you don't want to supply the db dump due to security reasons the schema should be fine. I'll generate some dummy date after importing everything and see if I can find the issue for your instance.

Thanks ahead of time for this!

**Reply by gallogabor** [reply]
Hi @ConorWebb96,

Today I updated Budibase from 2.7.11 to 2.7.19. In the changelog, I saw nothing about my problem, this was a simple casual update by me.

But, the above mentioned problem gone away after the update, I can insert and save JS code to the field, and the formatting works flawlessly. If I go back to JS editor, my inserted code shown and I can edit it, too.

Nevertheless, I can send the app export and the database schema (there are so much personal information in the database, I can not share those), if you wish to look at.

But, I have some other (more serious) problems than date formatting, and I would like to ask that may I open new discussion(s) about those topics/problems? I cannot find answers in issues or discussions.

I created this MySQL database back in 2021 to track my girlfriend's fitness class participants, and the plan was that I will use Budibase to produce a frontend which she can use for administration.
But at that time, I cannot make a working app with it (I think the version number was only a serial at that early time). So I filled in the data into the database by hand with DBeaver desktop app (making it somewhat easier with SQL triggers and views for reports). But she cannot do this, and cannot search the data already there.
Recently I checked on Budibase again to resolve the "problem" (it ran on my PC on Docker in the background for 2 years...) and found out that it has evolved so much. So let's try again. I dropped the old version, updated the stack with the current, and started a new app from scratch to eliminate the possible app upgrade problems.
But I ran into nearly the same problems and walls as 2 years earlier. I cannot make table relations to work right for me, and I cannot update tables with relations in them.

How can I get help about this?

**Reply by ConorWebb96** [reply]
No worries, happy it resolved itself. I recommend creating another discussion. Make sure to provide as much detail as possible including screenshots were applicable.

Thanks in advance!
